Add support for existing BigQuery Schema (replace #40)

abroglesc opened the pull request:
This takes the code proposed in PR #40 and attempts to address all outstanding items of concern as they were reviewed by @bxparks. #40 is very stale at this time and it does not seem that will get implemented at any point. I would like to request we close at #40 in favor of this PR but we should use that as a base of what changes I made since you last reviewed that PR.

The changes I have made as requested are:
1. Extend testdata structure to enable interpreting the existing schema as a string
    commit id: 8fbe52d0210867ec319c4a47fb9c9cd0d1ef1026
1. Convert added pytest tests to normal unittest (replaced fixtures with nested for loops which is worse but satisfies not using a different test suite)
    commit id: 1e286f14269066ce55628e143790a05f0a5a7b38
1. Delete pytest dep, and delete tox
    commit id: 1e286f14269066ce55628e143790a05f0a5a7b38
1. Add --existing_schema command line flag for reading in existing schema from a file. (super easy)
    commit id: 7afdf419a2c4a61c3b356d1b59ad5baaed8cc171
1. Write some tests that include an existing schema in testdata.txt
    commit id: fff6f5bf61a69edf67a1d19106dc5f42ef71cab4
1. Remove ERRORS INFORMED from the data_reader. This didn't make sense to me and I did not see the value. All errors should be part of the `error_logs` that are returned.
    commit id: fff6f5bf61a69edf67a1d19106dc5f42ef71cab4
1. Remove `type_mismatch_callback`. This was not properly tested and I didn't feel like maintaining/testing it.
    commit id: c5a57de54515e01de649f2705000a358f9688355
1. Fix bug as it related to bigquery not providing a mode for some nested JSON fields by defaulting to NULLABLE
    commit id: 96ca4ae148f6aa899665222d1c0e5a96968ecaae
1. Not requested but while updating the README I added the bash syntax flag to the `generate-schema` command examples


*Summarizing original changes from @bozzzzo.*
1. Break sanitize_names into it's own function and call this earlier in the schema generation process, not just during flattening.
1. Add ability to begin from an existing bigquery table schema by being able to reconstitute a schema_map from an existing bigquery schema.

abroglesc commented:
@bxparks This is ready for review. Happy to have your feedback and incorporate anything else that is needed to get this merged. I know this is a big change but it's a particularly useful one when you are loading data into existing tables.

bozzzzo commented:
@abroglesc Thanks for picking this up!

bxparks commented on tests/data_reader.py:
```diff
@@ -174,11 +221,11 @@ def read_errors_section(self):
-            return []
-        (tag, _) = self.parse_tag_line(tag_line)
+            return None, None
+        (tag, error_flags) = self.parse_tag_line(tag_line)
```
A bit confused why `error_flags` is extracted, instead of just being ignored as was originally. It does not seem like there were any flags added to the `ERRORS` section, so this flag is never used. The consequence is slightly more complicated return values and more complicated handling in the calling code.

abroglesc replied:
reverted in be3a37a

bxparks commented on tests/data_reader.py:
```diff
@@ -111,22 +120,31 @@ def read_chunk(self):
         Returns None if there are no more test chunks.
         """
-        data_flags, records = self.read_data_section()
+        data_flags, records, line = self.read_data_section()
```
nit: same comment as below, maybe rename `line` to `lineno` to avoid confusion?

abroglesc replied:
This has been implemented in 411f402

bxparks commented on tests/data_reader.py:
```diff
@@ -216,8 +264,8 @@ def read_schema_section(self):
             if tag in self.TAG_TOKENS:
-                if tag == 'SCHEMA':
-                    raise Exception("Unexpected SCHEMA tag")
+                if tag in ('DATA', 'ERRORS', 'EXISTING_SCHEMA', 'SCHEMA'):
```
Hmm, it's getting harder to keep track of which tags are allowed in which sections. Originally, the order of the tags were just: DATA, [ERRORS], SCHEMA, END. But now it's DATA, [EXISTING_SCHEMA], [ERRORS], SCHEMA, END. A better way would be to allow these sections to appears in any order. But that's a bit out of scope for this PR. If I get motivated, maybe I'll take a crack at it after merging in this PR... but realistically, it will probably not rise high enough on my priority list with so many other things going on. Too bad. At least this tidbit is recorded here.

abroglesc replied:
Created #59 to keep that nice to have visible after we merge this PR.

bxparks commented on tests/test_generate_schema.py:
```diff
@@ -495,15 +514,84 @@ def assert_error_messages(self, expected_error_map, error_logs):
+                                      " failed for case where: bq_entry={}\n"
+                                      "schema_generator created with values:"
+                                      "{}-{}-{}-{}"
+                                      .format(bq_entry,
```
nit: Can you change these into f-strings? I prefer them over the old formatters, and since I now require >=3.6, I can assume that f-strings are available.

abroglesc replied:
Addressed in fae9581

bxparks reviewed (commented):
This is fantastic, thanks for doing this. I tell you, no matter how many recursive functions I write and review, they still make my brain hurt trying to figure out if they are doing the right thing. Most of this looks great, I just have some small nits here and there. The major discussion point is inside `merge_schema_entry()` where you added some code to handle the situation of a field going from REQUIRED -> NULLABLE due to an existing schema. Please take a look at my notes and proposal, and let me know what you think.

abroglesc commented on tests/test_generate_schema.py:
```diff
@@ -434,23 +436,35 @@ class TestFromDataFile(unittest.TestCase):
-                self.verify_data_chunk(chunk_count, chunk)
-
-    def verify_data_chunk(self, chunk_count, chunk):
+                yield chunk
```
I honestly just took what bozozo was doing in the other PR and he had made this change. It doesn't seem like a bad change to me though. If you have a strong preference towards the old way I can change it back to that.

bxparks replied:
Ah, I see. My preference is to use the simplest techniques to do things, unless a more complicated solution is necessary. In this case, the old code was simpler and did the job, and using a generator here adds a tiny bit of cognitive overhead that is not really necessary. So I would prefer my original code.

abroglesc replied:
Reverted to old iteration way in d41556e

abroglesc commented on bigquery_schema_generator/generate_schema.py:
```diff
@@ -852,6 +939,9 @@ def main():
+    existing_schema_map = read_existing_schema_from_file(
+        args.existing_schema_path)
+    generator.run(schema_map=existing_schema_map)
     generator.run()
```
ack.

abroglesc commented on tests/test_generate_schema.py:
```diff
@@ -495,15 +514,84 @@ def assert_error_messages(self, expected_error_map, error_logs):
+                ('type', type),
+            ])
+
+    def validate_existing_schema_to_schema_map_entry(self,
```
ack, remnant of original tests by bozozo

abroglesc commented on bigquery_schema_generator/generate_schema.py:
```diff
@@ -778,17 +792,84 @@ def flatten_schema_map(
+def bq_schema_to_map(schema):
+    """ convert BQ JSON table schema representation to SchemaGenerator
+        schema_map representaton """
+    if isinstance(schema, dict):
```
ack

ZebraCat commented:
When is this expected to be merged and in the stable release?
we are working with big query in an environment where the schema can change every day, and this solution is perfect for us.

abroglesc commented:
Trying to get it done this week. Currently working through some other bugs related to how I am using the library then will focus back to getting this PR polished up (Since right now I am just vendoring this branch in my project to have it work)

abroglesc commented:
One part that I don't actually like currently about this change is the aggression in converting to lower for the names, I actually think I am going to see if I can resolve name collisions without forcing to lower for the final schema. I don't like that we are now changing the casing on new schema elements that are inserted. (e.g. `eventName` becomes `eventname` when run through this now).

We previously didn't have any tests which included capital letters in the source data or schema. We should not be changing the keys of the incoming source data (unless there is a collision, then we should use the original/existing name)

bxparks commented:
@abroglesc: With regards to the handling of the case, I agree we should be case-preserving, not case-insensitive, so we shouldn't be blindly lower casing everything. I think the solution is in the `schema_map` dictionary. The field `name` is stored in 2 places, the `schema_map.key` and `schema_map.info.name`. I think if we store the lower cased version in the dict key, and preserve the case in info.name, we can preserve the case.

bxparks commented:
@ZebraCat: If you need this right away, you can clone this repo, create a branch, and do a `git merge --squash` this branch into your branch and use the code from there.

@abroglesc: FYI, I'm eventually going to squash merge this into `develop`, instead of creating a merge commit, to reduce the huge amount of clutter in the git history. I don't normally use squash merge on GitHub, so I don't what that does to the authorship info, which I would like to preserve. I would have preferred the commits from @bozzzzo to be rebased off the `develop` branch, then have your changes added on top of that. But when I tried that, there were so many merge conflicts, it wasn't worth the effort.

abroglesc commented:
Ahh apologies, @bxparks I don't think github lends itself nicely to multiple people contributing to a PR unless they are given write permissions on the original repository so that has caused some issues here as it relates to how to continue with @bozzzzo's PR.

As for the using the key and the name, that sounds like a good plan.

abroglesc commented:
Fixed the casing issue. Added a couple tests for that as well. Just need to address the rest of the comments. Overall the logic fixes are mostly implemented just small changes left.

bxparks commented:
Hi, Just want to check in to get a status on this PR. The sooner we get this in the better, because it takes a fair bit of effort to pull in all the context info into my head to allow me to review this code properly. I don't work on this code frequently, so if we wait too long, I start forgetting things. 

BTW, there's a `Makefile` checked in at the root. You can run the unit tests and check flake8 errors by typing `make` at the top. You don't have to wait until you push tot GitHub.

abroglesc commented:
Hi @bxparks,

I will try to have this all addressed today. I had a bit of time off recently where I disconnected and just getting back into the swing of things.

abroglesc commented on bigquery_schema_generator/generate_schema.py:
```diff
@@ -348,11 +355,20 @@ def merge_schema_entry(
-                f'old=({old_status},{full_old_name},{old_mode},{old_type}); '
-                f'new=({new_status},{full_new_name},{new_mode},{new_type})')
-            return None
+            # primitive-types are conservatively deduced NULLABLE. In case we
```
Hi @bxparks this came from @bozzzzo's initial implementation.

I implemented your suggestion and then when writing a unit testcase to go from REQUIRED --> NULLABLE so I could ensure errors existed properly I could never trigger this. After a lot of tracing this is because when we deduce an existing schema we assume that status is hard since we can't change a type on line 286
```python
        # new 'soft' does not clobber old 'hard'
        if old_status == 'hard' and new_status == 'soft':
            return old_schema_entry
```


Given we assume all are hard and if we find a NULL field in a schema_entry we instantiate it with a status of `soft` we will never reach the condition to relax from REQUIRED --> NULLABLE.

Code for instantiating a NULL schema_entry https://github.com/bxparks/bigquery-schema-generator/blob/7d1b4ce892a174bc9dfaa6b5b3220691919f424f/bigquery_schema_generator/generate_schema.py#L427-L436


Should we move the logic about relaxing a field from REQUIRED --> NULLABLE or NULLABLE --> REQUIRED further up in this function or should we delete this status check and use it only when doing type conversions?

abroglesc replied:
I broke out the mode determination into it's own function and I call that within this soft/hard check and lower in the code where this comment initially referenced and it _seems_ to be working. Expect a commit with the change and updated testcase soon.

abroglesc replied:
This has now been implemented with tests and pushed in bb5745c8b55e7d4b51a191feab995bc0cb308c81

abroglesc commented:
@bxparks re: using Makefile for checking Flake8 I started doing that but it wasn't scanning `tests/` folder whereas the CI/CD in github was. This was added to the task in commit 9829bb0

abroglesc commented on tests/test_generate_schema.py:
```diff
@@ -434,23 +436,35 @@ class TestFromDataFile(unittest.TestCase):
+
+    def verify_data_chunk(self, chunk):
+        chunk_count = chunk['chunk_count']
+        line = chunk['line']
```
This has been implemented in 411f402

abroglesc commented:
@bxparks At this time I believe I have addressed all of your comments. Let me know if there is anything else we need to get this PR merged :)

abroglesc commented on bigquery_schema_generator/generate_schema.py:
```diff
@@ -778,17 +792,84 @@ def flatten_schema_map(
+    RECORD'''.split())
+
+BQ_TYPE_ALIASES = {
+    'INT64': 'INTEGER',
```
I realized I neglected to address this comment. Will add the test tonight.

As for #2, I believe it is showing legacy SQL types (at least on the table I tested with but I cannot find any documentation that states what it _should_ return). That said, I think coding this defensively like we did in anticipation of them potentially changing `bq show --schema` to start returning the new types is best practice here.

abroglesc replied:
A testcase has now been added that has these standard SQL types in the EXISTING_SCHEMA section of the testcase and expects legacy SQL types to be the generated schema.

abroglesc commented:
I now truly believe this is ready for a final review after addressing the one final comment I missed.

bxparks commented on bigquery_schema_generator/generate_schema.py:
```diff
@@ -360,37 +362,75 @@ def merge_schema_entry(
+            new_info['type'] = candidate_type
+        return new_schema_entry
+
+    def merge_mode(self, old_schema_entry, new_schema_entry, base_path):
```
Ok, this section is the crux of this PR, but unfortunately I am having an incredibly difficult time reviewing it. The GitHub code review tool is so clunky and difficult compared to what I'm used to. And the `git diff` output is completely messed up because the previous PR was not rebased off the current `develop` branch, so it's giving me diffs with lines that don't exist anymore, and I can't make any sense of it.

**Request**: Can you squash merge all your commits and Bozo's commits into a single commit, and send me another PR? I don't how if GitHub allows you to push those changes into this PR, or whether it will create another PR. I suspect the latter. This way, all your changes will be rebased off the current state of the code, so I can figure out what actually changed. Right now, I just cannot make any sense of it...

I think what you need to do is create a new branch, then do something like:
```
$ git co develop
$ git pull # make sure that you are synced with the repo
$ git co -b {new_branch}
$ git merge --squash abroglesc-existing-schema
$ git push
```
Then create a PR request on GitHub from `{new_branch}`.

abroglesc replied:
@bxparks created #61 as requested.

bxparks commented on bigquery_schema_generator/generate_schema.py:
```diff
@@ -105,17 +105,6 @@ def __init__(
-        # impossible, but more effort than I want to expend right now) to figure
-        # out which fields are missing so that we can mark the appropriate
-        # schema entries with 'filled=False'.
-        if infer_mode and input_format != 'csv':
```
I think this might be a bad idea. I know that you tried to recover some of this functionality in `flatten_schema_map()`, I am not sure that's sufficient. The reason is that if --infer_mode is enabled in JSON mode, it will do the wrong thing when JSON lines are missing those fields completely. All of the records will have a filled=True state, so I think the program will incorrectly infer the mode to be `REQUIRED`. This will break `bq load` when loading that dataset with this generated schema.

I realize that I was the one who suggested overloading the `--infer_mode` for the `REQUIRED->NULLABLE` transition, but I'm starting to think that we need a separate flag for this. To avoid enabling `--infer_mode` for JSON format.

abroglesc replied:
Will continue on the next PR I submit with squashed commits but why would this be bad given that added one more check to only apply this in `flatten_schema_map()` if the input_format is a CSV. If it's a JSON input format it will not elevate these to REQUIRED.

bxparks reviewed (commented):
I replied in one of my embedded comments that this PR has so many commits, and has Bozo's commit which are not rebased from `develop`, that I can no longer follow what it's doing. I added request in there, asking you to squash-merge all these changes into a single commit, rebased from `develop`, then send me another PR. Which will make it possible to get a sensible `git diff`.

bxparks commented:
Superceded by #61, closing.
